include/souper/Infer/Z3Driver.h

```diff
@@ -0,0 +1,256 @@
+      Solver.add(z3expr::sub_no_uoverflow(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+
+    if (I->K == Inst::Kind::MulNSW || I->K == Inst::Kind::MulNW) {
+      Solver.add(z3expr::mul_no_soverflow(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+    if (I->K == Inst::Kind::MulNUW || I->K == Inst::Kind::MulNW) {
+      Solver.add(z3expr::mul_no_uoverflow(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+
+    if (I->K == Inst::Kind::ShlNSW || I->K == Inst::Kind::ShlNW) {
+      Solver.add(z3expr::shl_no_soverflow(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+    if (I->K == Inst::Kind::ShlNUW || I->K == Inst::Kind::ShlNW) {
+      Solver.add(z3expr::shl_no_uoverflow(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+
+    if (I->K == Inst::Kind::UDivExact) {
+      Solver.add(z3expr::udiv_exact(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+    if (I->K == Inst::Kind::SDivExact) {
+      Solver.add(z3expr::sdiv_exact(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+
+    if (I->K == Inst::Kind::AShrExact) {
+      Solver.add(z3expr::ashr_exact(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+    if (I->K == Inst::Kind::LShrExact) {
+      Solver.add(z3expr::lshr_exact(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+  }
+
+  bool Translate(souper::Inst *I) {
+    if (!I) return true;
+    // unused translation; this is souper's internal instruction to represent overflow instructions
+    if (souper::Inst::isOverflowIntrinsicSub(I->K)) {
+      return true;
+    }
+
+    if (isCached(I)) {
+      return true;
+    }
+
+    auto Ops = I->Ops;
+    if (souper::Inst::isOverflowIntrinsicMain(I->K)) {
+      Ops = Ops[0]->Ops;
+    }
+
+    for (auto &&Op : Ops) {
+      if (!Translate(Op)) {
+        return false;
+      }
+      if (I->K == Inst::ExtractValue) {
+        break; // Break after translating main arg, idx is handled separately.
+      }
+
+    }
+
+    std::string Name = "";
+
+    if (NamesCache.find(I) != NamesCache.end()) {
+      Name = NamesCache[I];
+    } else if (I->Name != "") {
+      if (I->SynthesisConstID != 0) {
+        Name = "%" + souper::ReservedConstPrefix + std::to_string(I->SynthesisConstID);
+      } else {
+        Name = "%var_" + I->Name;
+      }
+    } else {
+      Name = "%" + std::to_string(InstNumbers++);
+    }
+    if (I->K == Inst::Var) {
+      NamesCache[I] = Name;
+    }
+
+    auto W = I->Width;
+    addExtraPreds(I);
+    switch (I->K) {
+      case souper::Inst::Var: {
+        Put(I, ctx.bv_const(Name.c_str(), W));
+        auto DFC = getDataflowConditions(I, IC);
+        if (DFC) AddConstraint(DFC);
+        return true;
+      }
+      case souper::Inst::Hole: {
+        llvm::report_fatal_error("Holes unimplemented in Z3Driver.");
+      }
+      case souper::Inst::Const: {
+        Put(I, ctx.bv_val(I->Val.toString(10, false).c_str(), W));
+        // inefficient?
+        return true;
+      }
+
+      case souper::Inst::Phi: {
+        if (Ops.size() != 1) {
+          assert(false && "Phi with muliple arguments unimplemented");
```

Comment:
plz fix this before we merge your patch

Reply:
Phi's are handled at another level of abstraction. Both this and AliveDriver are responsible for verifying a transformation, given a precondition. Another layer decomposes LHS's with 'actual' phi nodes into multiple such (Pre => LHS == RHS?) problems.

Reply:
ah, nice, thanks-- please add a comment saying phis have already been broken down before you get to this point